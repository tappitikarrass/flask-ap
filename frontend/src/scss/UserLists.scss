@use "base16" as *;

#user-lists.content {
    display: flex;
    flex-direction: column;
    gap: 1em;
}


// desktop
@media screen and (min-width: 810px) {
    #user-lists.content {
        width: 70vw;
        margin-left: 15vw;
    }
    #create-list-bt {
        width: 15vw;
    }
}

// tablet
@media screen and (min-width: 500px) and (max-width: 810px) {
    #user-lists.content {
        width: 100vw;
        margin-left: 0vw;
    }
    #create-list-bt {
        width: 15vw;
    }
}

// phone
@media screen and (max-width: 500px) {
    #user-lists.content {
        width: 100vw;
        margin-left: 0;
    }
}
