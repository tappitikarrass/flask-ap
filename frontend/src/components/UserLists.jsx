import React, { useRef, useState, useEffect } from 'react';
import { useCookies } from 'react-cookie';
import { useForm } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import '../scss/UserLists.scss';

export default function UserLists() {
  const navigate = useNavigate();
  const createListForm = useRef();
  const [cookies] = useCookies('token', 'username', 'user_id');
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();
  const [lists, setLists] = useState([]);

  useEffect(() => {
    async function effect() {
      const response = await fetch(
        `http://localhost/backend/list/${cookies.username}`,
        {
          method: 'GET',
          headers: {
            Authorization: `Bearer ${cookies.token}`,
          },
        },
      );
      const json = await response.json();
      setLists(json);
    }
    if (cookies.token === 'undefined') {
      navigate('/login');
    } else {
      effect();
    }
  }, []);

  async function onSubmit() {
    const listData = new FormData(createListForm.current);

    const response = await fetch(
      'http://localhost/backend/list',
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${cookies.token}`,
        },
        body: JSON.stringify(
          {
            user_id: cookies.user_id,
            name: listData.get('listname'),
          },
        ),
      },
    );
    const json = await response.json();
    window.location.reload();
    return json;
  }

  async function deleteList(listId) {
    const response = await fetch(
      `http://localhost/backend/list/${cookies.username}/${listId}`,
      {
        method: 'DELETE',
        headers: {
          Authorization: `Bearer ${cookies.token}`,
        },
      },
    );
    if (response.status === 200) {
      window.location.reload();
    }
  }

  return (
    <div id="user-lists" className="content">
      <form
        id="create-list"
        ref={createListForm}
        method="post"
        acceptCharset="utf-8"
        onSubmit={handleSubmit(onSubmit)}
      >
        <div className="ui-row">
          <input
            {...register('listname', {
              required: 'List name is required',
              minLength:
              {
                value: 5,
                message: 'Min list name length is 5 characters.',
              },
              maxLength:
              {
                value: 256,
                message: 'Max list name length is 256 characters.',
              },
            })}
            type="text"
            className="field"
            placeholder="List name"
          />
          <button
            id="create-list-bt"
            type="submit"
            className="bt-blue-end"
          >
            Create
          </button>
        </div>
      </form>
      <p>{errors.listName?.message}</p>
      <table id="lists-table">
        <tbody>
          <tr>
            <td>#</td>
            <td>Name</td>
            <td />
            <td />
          </tr>
          {
            lists.filter((list) => list.user_id === Number(cookies.user_id)).map((value) => (
              <tr key={value.list_id}>
                <td>{value.list_id}</td>
                <td>{value.name}</td>
                <td>
                  <button
                    id="delete-list-bt"
                    className="bt-green-end"
                    type="submit"
                    onClick={() => {
                      navigate(`/list/${value.list_id}`);
                    }}
                  >
                    Open
                  </button>
                </td>
                <td>
                  <button
                    id="delete-list-bt"
                    className="bt-red-end"
                    type="submit"
                    onClick={() => {
                      deleteList(value.list_id);
                    }}
                  >
                    Delete
                  </button>
                </td>
              </tr>
            ))
          }
        </tbody>
      </table>
    </div>
  );
}
