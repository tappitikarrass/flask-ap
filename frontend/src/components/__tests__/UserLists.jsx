import React from "react";
import { HashRouter } from "react-router-dom";
import { create, act } from "react-test-renderer";
import UserLists from '../UserLists';

describe('UserLists', () => {
  it('',() => {
    let component;
    act(() => {
      component = create(
        <HashRouter>
          <UserLists />
        </HashRouter>
      );
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});
