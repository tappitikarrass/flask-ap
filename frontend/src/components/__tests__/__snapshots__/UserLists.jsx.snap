// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserLists  1`] = `
<div
  className="content"
  id="user-lists"
>
  <form
    acceptCharset="utf-8"
    id="create-list"
    method="post"
    onSubmit={[Function]}
  >
    <div
      className="ui-row"
    >
      <input
        className="field"
        name="listname"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="List name"
        type="text"
      />
      <button
        className="bt-blue-end"
        id="create-list-bt"
        type="submit"
      >
        Create
      </button>
    </div>
  </form>
  <p />
  <table
    id="lists-table"
  >
    <tbody>
      <tr>
        <td>
          #
        </td>
        <td>
          Name
        </td>
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
